providers/repo.rb

```diff
@@ -138,17 +138,45 @@ def install_gem
   end
 end
 
-def read_token(repo_url, gems=false)
-  return repo_url unless new_resource.master_token
+### The hostname lookup chain is:
+### 1) node['packagecloud']['override_hostname']
+### 2) node['fqdn']
+### 3) node['hostname']
+### We persist this hostname result in node['packagecloud']['_stored_hostname']
+### between chef runs so that we can determine if a new read token should be issued
+### from the packagecloud API for new/changed hostnames. See #should_issue_read_token?
+def get_hostname
+  hostname = override_or_fqdn_or_hostname
+  if hostname.nil?
+    raise("Can't determine hostname!  Set node['packagecloud']['override_hostname'] " \
+          "if it cannot be automatically determined by Ohai.")
+  else
+    node.set['packagecloud']['_stored_hostname'] = hostname
+    hostname
+  end
+end
 
-  if node['fqdn'].nil?
-    Chef::Log.fatal("This node's fqdn is set to nil, so a read token cannot be issued!" \
-                    "Please change your fqdn settings.")
+## Legacy deployments who have node['packagecloud']['hostname'] set are
+## migrated over
+def stored_hostname
+  if node['packagecloud']['hostname']
+    node.set['packagecloud']['_stored_hostname'] = node['packagecloud']['hostname']
   end
+  node['packagecloud']['_stored_hostname']
+end
 
-  if node['packagecloud'][new_resource.repository].nil? ||
-     node['packagecloud']['hostname'] != node['fqdn']
+def should_issue_read_token?
```

Comment:
Monkeypatching the providers wouldn't help -- we return 404 if the repo doesn't exist and also in the case of an invalid read token. There's no way to differentiate those cases and `apt` and `yum` don't really work well enough to capture this data reliably.

Perhaps we can introduce a way to allow users to extract their readtokens from an [encrypted databag](https://docs.chef.io/data_bags.html#encrypt-a-data-bag-item). I'm not sure yet how, though, and it'd be a separate PR.

It's important to remember that these tokens will be written to the package manager's configuration file on the machine. We could encrypt them as far as chef is concerned, but they're still readable as a privileged user on the machine -- this is why these tokens don't allow you to do anything other than read packages and metadata.